runtime/src/stakes.rs

```diff
@@ -225,24 +225,58 @@ impl Stakes<StakeAccount> {
     /// full account state for respective stake pubkeys. get_account function
     /// should return the account at the respective slot where stakes where
     /// cached.
-    pub(crate) fn new<F>(stakes: &Stakes<Delegation>, get_account: F) -> Result<Self, Error>
+    pub(crate) fn new<F>(stakes: Stakes<Delegation>, get_account: F) -> Result<Self, Error>
     where
-        F: Fn(&Pubkey) -> Option<AccountSharedData>,
+        F: Fn(&Pubkey) -> Option<AccountSharedData> + Sync,
     {
-        let stake_delegations = stakes.stake_delegations.iter().map(|(pubkey, delegation)| {
-            let Some(stake_account) = get_account(pubkey) else {
-                return Err(Error::StakeAccountNotFound(*pubkey));
-            };
-            let stake_account = StakeAccount::try_from(stake_account)?;
-            // Sanity check that the delegation is consistent with what is
-            // stored in the account.
-            if stake_account.delegation() == *delegation {
-                Ok((*pubkey, stake_account))
-            } else {
-                Err(Error::InvalidDelegation(*pubkey))
-            }
-        });
+        // im::HashMap doesn't support rayon so we manually build a temporary vector. Note this is
+        // what std HashMap::par_iter() does internally too.
+        let stake_delegations_vec = stakes.stake_delegations.into_iter().collect::<Vec<_>>();
+        let stake_delegations = stake_delegations_vec
+            .into_par_iter()
+            // We use fold/reduce to aggregate the results, which does a bit more work than calling
+            // collect_vec_list() and then manually inserting into a map, but it does it in background
+            // threads, so effectively it's faster.
```

Comment:
isn't it just faster if the par-iter operations just return a `Vec` (i.e. a simple map operation) and we convert the `Vec` to `im::HashMap` at the end?

Reply:
That's how I originally wrote it, but it's slower.

collect() allocates the result vec in the caller thread. A map in place isn't possible conceptually - the layouts of Delegation and StakeAccount don't match, so a new vector needs to be allocated. Besides, rayon doesn't support in place map anyway: when you do `collect::<Vec<_>>()` it does `let mut v = Vec::new(); v.reserve(iter.len()); v.par_extend(iter)`. So there is the extra cost of a large Vec allocation in the main thread, but that's a relatively small factor.

The dominating factor is that `im::HashMap::from_iter(collected_vec.into_iter())` is very slow. By using fold/reduce, we build im::HashMap(s) in background threads - doing a bit more work merging intermediary maps - but we build the two halves of the map in parallel in background threads, until we do a final reduce which needs to do len / 2 inserts. In practice on my machine this is 2-3 seconds faster.